File: mrgeo-mapalgebra/mrgeo-mapalgebra-terrain/src/main/java/org/mrgeo/mapalgebra/AspectMapOp.java
/*
 * Copyright 2009-2015 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 */

package org.mrgeo.mapalgebra;


import org.mrgeo.data.DataProviderFactory;
import org.mrgeo.data.image.MrsImageDataProvider;
import org.mrgeo.mapalgebra.old.MapOpHadoop;
import org.mrgeo.mapalgebra.old.ParserAdapterHadoop;
import org.mrgeo.mapalgebra.old.RasterMapOpHadoop;
import org.mrgeo.mapalgebra.parser.ParserNode;
import org.mrgeo.mapreduce.job.JobCancelledException;
import org.mrgeo.mapreduce.job.JobFailedException;
import org.mrgeo.opimage.MrsPyramidDescriptor;
import org.mrgeo.progress.Progress;
import org.mrgeo.spark.AspectDriver;

import java.io.IOException;
import java.util.HashSet;
import java.util.Set;
import java.util.Vector;

public class AspectMapOp extends RasterMapOpHadoop implements InputsCalculator
{
String units = "rad";

public static String[] register()
{
  return new String[] { "aspect" };
}

@Override
public void addInput(MapOpHadoop n) throws IllegalArgumentException
{
  if (!(n instanceof RasterMapOpHadoop))
  {
    throw new IllegalArgumentException("Can only run aspect() on raster inputs");
  }
  if (_inputs.size() >= 1)
  {
    throw new IllegalArgumentException("Can only run aspect() on a single raster input");
  }

  _inputs.add(n);
}

@Override
public void build(Progress p) throws IOException, JobFailedException, JobCancelledException
{
  p.starting();
  // check that we haven't already calculated ourselves
  if (_output == null)
  {

    String input = ((RasterMapOpHadoop) _inputs.get(0)).getOutputName();

    AspectDriver.aspect(input, units, getOutputName(), createConfiguration());

    MrsImageDataProvider dp = DataProviderFactory.getMrsImageDataProvider(getOutputName(),
        DataProviderFactory.AccessMode.READ, getProviderProperties());
    _output = MrsPyramidDescriptor.create(dp);
  }
  p.complete();
}

@Override
public Vector<ParserNode> processChildren(Vector<ParserNode> children, ParserAdapterHadoop parser)
{
  Vector<ParserNode> result = new Vector<ParserNode>();

  if (children.size() > 2)
  {
    throw new IllegalArgumentException(
        "Aspect takes one or two arguments. single-band raster elevation and optional unit format (\"deg\" or \"rad\")");
  }

  result.add(children.get(0));

  if (children.size() == 2)
  {
    String units = MapOpHadoop.parseChildString(children.get(1), "units", parser);
    if (!(units.equalsIgnoreCase("deg") || units.equalsIgnoreCase("rad")))
    {
      throw new IllegalArgumentException("Units must be \"deg\", or \"rad\".");
    }
    this.units = units;
  }

  return result;
}
@Override
public Set<String> calculateInputs()
{
  Set<String> inputPyramids = new HashSet<String>();
  if (_outputName != null)
  {
    inputPyramids.add(_outputName);
  }
  return inputPyramids;
}


@Override
public String toString()
{
  return "aspect()";
}
}

File: mrgeo-mapalgebra/mrgeo-mapalgebra-terrain/src/main/java/org/mrgeo/mapalgebra/AspectMapOp.scala
/*
 * Copyright 2009-2016 DigitalGlobe, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and limitations under the License.
 *
 */

package org.mrgeo.mapalgebra

import java.io.{ObjectInput, ObjectOutput}

import org.mrgeo.mapalgebra.parser.{ParserException, ParserNode}
import org.mrgeo.mapalgebra.raster.RasterMapOp
import org.mrgeo.utils.FloatUtils

object AspectMapOp extends MapOpRegistrar {
  override def register:Array[String] = {
    Array[String]("aspect")
  }

  def create(raster:RasterMapOp, units:String = "rad", flatValue:Double = -1.0):MapOp = {
    new AspectMapOp(Some(raster), units, flatValue)
  }

  override def apply(node:ParserNode, variables:String => Option[ParserNode]):MapOp =
    new AspectMapOp(node, variables)
}

class AspectMapOp extends SlopeAspectMapOp {
  final val TWO_PI:Double = 2 * Math.PI
  final val THREE_PI_OVER_2:Double = (3.0 * Math.PI) / 2.0

  var flatValue:Double = -1.0

  override def computeTheta(normal:(Double, Double, Double)):Double = {
    // if the z component of the normal is 1.0, the cell is flat, so the aspect is undefined.
    // For now, we'llset it to 0.0, but another value could be more appropriate.
    if (FloatUtils.isEqual(normal._3, 1.0)) {
      flatValue
    }
    else {
      // change from (-Pi to Pi) to ( [0 to 2Pi) ), make 0 deg north (+ 3pi/2)
      // convert to clockwise
      val t = TWO_PI - (Math.atan2(normal._2, normal._1) + THREE_PI_OVER_2) % TWO_PI
      if (FloatUtils.isEqual(t, TWO_PI)) {
        flatValue
      }
      else {
        t
      }
    }
  }

  override def readExternal(in:ObjectInput):Unit = {
    super.readExternal(in)
    flatValue = in.readDouble()
  }

  override def writeExternal(out:ObjectOutput):Unit = {
    super.writeExternal(out)
    out.writeDouble(flatValue)
  }

  private[mapalgebra] def this(inputMapOp:Option[RasterMapOp], units:String, flatValue:Double) = {
    this()

    initialize(inputMapOp, units)
    this.flatValue = flatValue
  }

  private[mapalgebra] def this(node:ParserNode, variables:String => Option[ParserNode]) = {
    this()

    if (node.getNumChildren < 1) {
      throw new ParserException(node.getName + " requires at least one argument")
    }
    else if (node.getNumChildren > 4) {
      throw new ParserException(node.getName + " takes no more than four arguments")
    }

    val inputMapOp = RasterMapOp.decodeToRaster(node.getChild(0), variables)
    var units:String = "rad"
    if (node.getNumChildren >= 2) {
      units = MapOp.decodeString(node.getChild(1)) match {
        case Some(s) => s
        case _ => throw new ParserException("Error decoding string")
      }
    }
    initialize(inputMapOp, units)
    if (node.getNumChildren >= 3) {
      flatValue = MapOp.decodeDouble(node.getChild(2)) match {
        case Some(d) => d
        case _ => throw new ParserException("Error decoding double from " + node.getChild(2).getName)
      }
    }
  }
}
